parser2/src/grammar.lalrpop

```diff
@@ -0,0 +1,233 @@
+use std::iter;
+
+use {Expr, Literal, Ident, MatchArm, Pattern, Type, TypeBinding, ValueBinding};
+use tokenize::{Error, /*Location, */ Token};
+
+grammar<'input>(src: &'input str);
+
+extern {
+    // type Location = Location;
+    type Location = usize;
+    type Error = Error;
+
+    enum Token<'input> {
+        "and" => Token::And,
+        "else" => Token::Else,
+        "if" => Token::If,
+        "in" => Token::In,
+        "let" => Token::Let,
+        "match" => Token::Match,
+        "then" => Token::Then,
+        "type" => Token::Type,
+        "with" => Token::With,
+
+        "Name" => Token::Name(<&'input str>),
+        "Operator" => Token::Operator(<&'input str>),
+        "ByteLiteral" => Token::ByteLiteral(<u8>),
+        "IntLiteral" => Token::IntLiteral(<i64>),
+        "FloatLiteral" => Token::FloatLiteral(<f64>),
+        "StringLiteral" => Token::StringLiteral(<String>),
+        "CharLiteral" => Token::CharLiteral(<char>),
+
+        ":" => Token::Colon,
+        "," => Token::Comma,
+        "." => Token::Dot,
+        "=" => Token::Equals,
+        "\"" => Token::Lambda,
+        "{" => Token::LBrace,
+        "[" => Token::LBracket,
+        "(" => Token::LParen,
+        "|" => Token::Pipe,
+        "->" => Token::RArrow,
+        "}" => Token::RBrace,
+        "]" => Token::RBracket,
+        ")" => Token::RParen,
+
+        "OpenBlock" => Token::OpenBlock,
+        "CloseBlock" => Token::CloseBlock,
+        "Separator" => Token::Separator,
+    }
+}
+
+// Utils
+
+Comma<Rule>: Vec<Rule> =
+    <rules: (<Rule> ",")*> <last: Rule?> => {
+        let mut rules = rules;
+        rules.extend(last);
+        rules
+    };
+
+And<Rule>: Vec<Rule> =
+    <rules: (<Rule> "and")*> <last: Rule> =>
+        rules.into_iter().chain(iter::once(last)).collect();
```

Comment:
Avoid allocating a new `Vec` by pushing the last rule instead (as you do in the Comma macro).